src/phasorpy/phasor.py

```diff
@@ -3070,11 +3110,157 @@ def _median(
     return numpy.nanmedian(real, **kwargs), numpy.nanmedian(imag, **kwargs)
 
 
-def _median_filter(
-    real: ArrayLike,
-    imag: ArrayLike,
+def _median_filter_2d(
+    real: NDArray[Any],
+    imag: NDArray[Any],
+    axes: Sequence[int],
+    /,
+    *,
     repeat: int = 1,
-    size: int | tuple[int] | None = 3,
+    size: int = 3,
+    num_threads: int | None = None,
+) -> tuple[NDArray[Any], NDArray[Any]]:
+    """Return the phasor coordinates after applying a 2D median filter.
+
+    Parameters
+    ----------
+    real : ndarray
+        Real components of the phasor coordinates.
+    imag : ndarray
+        Imaginary components of the phasor coordinates.
+    axes : sequence of int
+        Axes along which to apply the filter.
+    repeat : int, optional
+        Number of times to apply filter. The default is 1.
+    size : int, optional
+        The size of the median filter kernel. Default is 3.
+    num_threads : int, optional
+        Number of OpenMP threads to use for parallelization.
+        By default, multi-threading is disabled.
+        If zero, up to half of logical CPUs are used.
+        OpenMP may not be available on all platforms.
+
+    Returns
+    -------
+    real : ndarray
+        Filtered real component of phasor coordinates.
+    imag : ndarray
+        Filtered imaginary component of phasor coordinates.
+
+    """
+    real = numpy.asarray(real)
+    if real.dtype == numpy.float32:
+        real = real.copy()
+    else:
+        real = real.astype(float)
+
+    imag = numpy.asarray(imag)
+    if imag.dtype == numpy.float32:
+        imag = imag.copy()
+    else:
+        imag = imag.astype(float)
+
+    if len(axes) != 2:
+        return _median_filter_nd(real, imag, axes, repeat=repeat, size=size)
+
+    num_threads = number_threads(num_threads)
+
+    for _ in range(repeat):
+        for index in numpy.ndindex(
+            *[real.shape[ax] for ax in range(real.ndim) if ax not in axes]
+        ):
+            index_list: list[int | slice] = list(index)
+            for ax in axes:
+                index_list = index_list[:ax] + [slice(None)] + index_list[ax:]
+            full_index = tuple(index_list)
+
+            real_slice = real[full_index]
+            imag_slice = imag[full_index]
+
+            filtered_real_slice = numpy.empty_like(real_slice)
+            filtered_imag_slice = numpy.empty_like(imag_slice)
+
+            _apply_2d_median_filter(
+                real_slice, filtered_real_slice, size, num_threads
+            )
+            _apply_2d_median_filter(
+                imag_slice, filtered_imag_slice, size, num_threads
+            )
+
+            real[full_index] = filtered_real_slice
+            imag[full_index] = filtered_imag_slice
```

Comment:
No need to allocate two filtered slices or index again on copy:
```Python
            filtered_slice = numpy.empty_like(real_slice)

            _apply_2d_median_filter(
                real_slice, filtered_slice, size, num_threads
            )
            numpy.copyto(real_slice, filtered_slice)

            _apply_2d_median_filter(
                imag_slice, filtered_slice, size, num_threads
            )
            numpy.copyto(imag_slice, filtered_slice)
```

The `numpy.copyto()` function call could be avoided by moving the copy into the `_apply_2d_median_filter` function.

Also, the allocation of the filtered slice can probably be moved out of the loops. I don't think the slice shape or dtype changes.

Reply:
Thanks for noticing this Christoph, I think it makes sense to make this changes. It didn't have that much of an impact in execution times, but copying in the `_apply_2d_median_filter` function has the advantage of using parallel processing which can further improve performance.

Also at first I though of moving the indexing also out of the loop: 

```Python
for index in numpy.ndindex(
    *[real.shape[ax] for ax in range(real.ndim) if ax not in axes]
):
    index_list: list[int | slice] = list(index)
    for ax in axes:
        index_list = index_list[:ax] + [slice(None)] + index_list[ax:]
    full_index = tuple(index_list)
```

But this can be a problem for large images since the indices should be stored in some form of sequence with all values  not in `axes`, so this can be quite large. And I don't think the performance improvement would be significant. But if you think it makes sense I can make the changes.

Reply:
OK